### PR TITLE
Chain history data

### PR DESCRIPTION
**Pull request by lgobbi-atix**
# Context

As a developer wishing to access a full chain history
I want to have the option of using a remote service without depending on a specific one,
so that I can write application logic without being coupled to a particular service provider

# Proposed Solution

- Implement `ChainHistoryProvider`
- Implement `ChainHistoryHttpService`
- Implement `ChainHistoryHttpProvider`
- Implement `ChainHistoryDbSyncProvider`
- Implement `ChainHistoryBuilder`

# Important Changes Introduced
- Added ChainHistory provider, queries builder and http service
- Added DBSyncChainHistory
- query transaction by addresses or hashes
- query blocks by hashes

### CODE REVIEW

File: packages/cardano-services/src/ChainHistory/openApi.json
Comment (lgobbi-atix) [inline packages/cardano-services/src/ChainHistory/openApi.json:397]: Not sure about this

File: packages/cardano-services/src/ChainHistory/ChainHistoryHttpService.ts
Comment (lgobbi-atix) [inline packages/cardano-services/src/ChainHistory/ChainHistoryHttpService.ts:46]: Is there any reason why we should pass args and response types manually? Does it make sense to change `providerHandler` to be able to use it like this instead?

```ts
providerHandler<typeof chainHistoryProvider.blocksByHashes>(async (args, _, res, handler) => {
``` 

Or even something like this: 
```ts
providerHandler(chainHistoryProvider.blocksByHashes)(async (args, _, res, handler) => {
        try {
          return HttpServer.sendJSON(res, await handler(...args));
        } catch (error) {
          logger.error(error);
          return HttpServer.sendJSON(res, new ProviderError(ProviderFailure.Unhealthy, error), 500);
        }
      }
```

Comment (mkazlauskas) [reply packages/cardano-services/src/ChainHistory/ChainHistoryHttpService.ts:46]: No reason, I agree, inferring the types would be better

File: packages/core/src/Asset/util/assetId.ts
Comment (mkazlauskas) [inline packages/core/src/Asset/util/assetId.ts:17]: Can't we do just `policyId.toString() + assetName.toString()`? Looking at the test for this it seems to be the case

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/cardano-services/src/ChainHistory/openApi.json
Comment (rhyslbw) [reply packages/cardano-services/src/ChainHistory/openApi.json:397]: Specifically?:
```
              "oneOf": [{ "type": "array", "items": { "$ref": "#/components/schemas/Address" } }, { "type": "number" }]
```

Yeah, why is there a `number` expected?

Comment (lgobbi-atix) [reply packages/cardano-services/src/ChainHistory/openApi.json:397]: The number is because `transactionsByAddresses` has a `sinceBlock` optional argument. https://github.com/input-output-hk/cardano-js-sdk/pull/243/files#diff-cc6fbcd6f5f3a6dee2c12a90216ba559ea451cb187699b838f8bed6b600d65c3R339

If we base this on that method arguments like the others the request body should be something like `args: [[addresses], blockNo]`. But I don't think we can validate that with just open api, what I did here would allow `[array, number]`, `[number,array]`, `[array,array]`, `[number, number]`, etc. We could still validate them manually in the provider but the endpoint documentation would be misleading.

Should I change it to an object instead: `args: [{ addresses, blockNo? }]`?
It will be different than the others though as the request body and the provider method arguments don't match. And if you omit `blockNo` the request would look like `args: [{ addresses: Address[] }]`, unlike the others where it's just `args: [[hashes]]`, which is a bit inconsistent I guess

Comment (rhyslbw) [reply packages/cardano-services/src/ChainHistory/openApi.json:397]: > We could still validate them manually in the provider handler but the endpoint documentation would be misleading

Which is an expression of the API :-)  Let's aim for self-describing interfaces at all times 👍 
 
> Should I change it to an object instead: args: [{ addresses, blockNo? }]?

With a slight consistency correction: 
`args: [{ addresses, sinceBlock? }]`

> It will be different than the others though as the request body and the provider method arguments don't match.
...
 which is a bit inconsistent I guess
 
 Please log it as a SPIKE for @mkazlauskas to review, assigned to the production-readiness epic, and we'll press on here.

File: packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts
Comment (iadmytro) [inline packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts:169]: missing cases for:
-    MirCertificate
-    GenesisKeyDelegationCertificate  
-    StakeKeyDeregistrationCertificate

**Review by iadmytro — CHANGES_REQUESTED** [review]

File: packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts
Comment (lgobbi-atix) [reply packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts:193]: I'm not sure if we can get `GenesisKeyDelegationCertificate` from db-sync, I didn't make any queries to get them and they are not being fetched by the blockfrost provider either.

Added the others in https://github.com/input-output-hk/cardano-js-sdk/pull/243/commits/5372e875ce7ce0f93277b75e1f5b2bdbaa008274

File: packages/cardano-services/src/ChainHistory/openApi.json
Comment (lgobbi-atix) [reply packages/cardano-services/src/ChainHistory/openApi.json]: I changed the open api specification to an object in this commit https://github.com/input-output-hk/cardano-js-sdk/pull/243/commits/50827879ec95916cb72e72952de9214a3a35de68, but had to delete these tests here https://github.com/input-output-hk/cardano-js-sdk/pull/243/commits/50827879ec95916cb72e72952de9214a3a35de68#diff-cc6fbcd6f5f3a6dee2c12a90216ba559ea451cb187699b838f8bed6b600d65c3L276 because the open api validation fails when the http provider method makes a request to the http service as the request and method arguments are different now (request: `({ addresses, sinceBlock })`, method: `(addresses, sinceBlock)`). 

Besides that, I don't really see the point of those kinds of tests, the provider already has `chainHistoryHttpProvider.test.ts` and the HttpService is already covered by the tests that use `doChainHistoryRequest`

So I'm not sure what to do, should we revert this commit and leave the request body as an array? Should we change the provider method arguments? Should we leave it like this?

Comment (rhyslbw) [reply packages/cardano-services/src/ChainHistory/openApi.json]: >  Should we change the provider method arguments to match the http service?

Yes. I'll make a ticket to review all our Provider interfaces for consistency

File: packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts
Comment (rhyslbw) [reply packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts:193]: Please leave a comment in the test to explain why there are no tests for these cert types @lgobbi-atix

Comment (iadmytro) [reply packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts:193]: It's blocked by https://github.com/input-output-hk/cardano-db-sync/issues/290

Comment (rhyslbw) [reply packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts:193]: https://github.com/input-output-hk/cardano-js-sdk/commit/07a978ee39e38f0f8a62a662895af610b22db10c#diff-3e16e6fa3a0c30bc7dc482538653e7edca22402ca67614ba46757d93f533b0e8R150

File: packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/types.ts
Comment (rhyslbw) [inline packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/types.ts:107]: Only types or interfaces should be in this file. Please move all functions to a new `./util.ts`

File: packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/types.ts
Comment (rhyslbw) [inline packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/types.ts:56]:
```suggestion
export interface TxOutputModel extends Cardano.TxOut {
```
?

File: packages/cardano-services/src/Program/loadHttpServer.ts
Comment (rhyslbw) [inline packages/cardano-services/src/Program/loadHttpServer.ts:63]: Should we hoist this before the switch to share the instance across services?

File: packages/cardano-services/src/types.ts
Comment (rhyslbw) [inline packages/cardano-services/src/types.ts:2]: This is generic, but probably not worth moving until we have implemented the `util` package.

File: packages/cardano-services/src/util/hexString.ts
Comment (rhyslbw) [inline packages/cardano-services/src/util/hexString.ts:1]: Also generic, but also probably not worth moving until we have implemented the util package.

File: packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap
Comment (rhyslbw) [inline packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap:143]: What changed to invalidate the snapshot?

**Review by rhyslbw — CHANGES_REQUESTED** [review]
Need to resolve the lint warning:
image

Looking at the code I'd be fine with that rule being disbled inline

File: packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.ts
Comment (iadmytro) [inline packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.ts:141]: better to rename it to poolRegisterCerts

File: packages/cardano-services-client/src/ChainHistoryProvider/chainHistoryHttpProvider.ts
Comment (mkazlauskas) [reply packages/cardano-services-client/src/ChainHistoryProvider/chainHistoryHttpProvider.ts:5]: I agree, it is cleaner and could make a potential future refactor of hoisting this to a new BlocksProvider easier. I'm fine with a temporary divergence, changing the existing paths is a small task anyway. `blocks/by-hashes` :+1:

File: packages/cardano-services/src/util/hexString.ts
Comment (mkazlauskas) [reply packages/cardano-services/src/util/hexString.ts:1]: Yep, similar to [bytesToHex](https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/core/src/util/misc/bytesToHex.ts) that's currently in `core`

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap
Comment (mkazlauskas) [reply packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap:143]: Snapshots failed [here](https://github.com/input-output-hk/cardano-js-sdk/runs/6507593263?check_suite_focus=true) too after merge, but the PR checks were green :thinking:

File: packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/mappers.ts
Comment (rhyslbw) [reply packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/mappers.ts:215]: or make it an optional property?

File: packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap
Comment (lgobbi-atix) [reply packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap:143]: I needed to add some tables to the dump and the database used before was a db that Juan Cruz had in his computer. The one I used had different ids (but same data) for some records, so I had to update the snapshots. We are now sharing the same one

File: packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/mappers.ts
Comment (rhyslbw) [reply packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/mappers.ts:215]: > I'd prefer not to mix the 2 styles (optional vs no property)

Agree. Let's drop it

File: packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap
Comment (rhyslbw) [reply packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap:143]: Can the original one be used instead @lgobbi-atix, or would that just shift the impact?

Comment (rhyslbw) [reply packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap:143]: Can we generate a snapshot from the original DB, or would that just shift the impact?

Comment (lgobbi-atix) [reply packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap:143]: Thinking about it with Juan Cruz maybe it's better to use a db with the [snapshots](https://updates-cardano-testnet.s3.amazonaws.com/cardano-db-sync/index.html#12/) that db-sync provides but without running db-sync.
That way all devs could have their own frozen local database (using the same snapshot version of course) and then run the dump script on it if they need to add more tables/blocks to the fixture.

There shouldn't be any issues with ids anymore if everyone is using the same snapshot and db-sync is not running.

File: packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/mappers.ts
Comment (lgobbi-atix) [reply packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/mappers.ts:215]: Should we do this in another task?

File: packages/core/src/Provider/ChainHistoryProvider/types.ts
Comment (lgobbi-atix) [reply packages/core/src/Provider/ChainHistoryProvider/types.ts]: Done in 70dd1e23cbb08257e051305e215dc03c001d0d99

File: packages/cardano-services-client/src/ChainHistoryProvider/chainHistoryHttpProvider.ts
Comment (lgobbi-atix) [reply packages/cardano-services-client/src/ChainHistoryProvider/chainHistoryHttpProvider.ts]: Done in 70dd1e23cbb08257e051305e215dc03c001d0d99

File: packages/cardano-services-client/test/ChainHistoryProvider/chainHistoryProvider.test.ts
Comment (lgobbi-atix) [reply packages/cardano-services-client/test/ChainHistoryProvider/chainHistoryProvider.test.ts]: Done in 70dd1e23cbb08257e051305e215dc03c001d0d99

File: packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/types.ts
Comment (lgobbi-atix) [reply packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/types.ts]: Done in 70dd1e23cbb08257e051305e215dc03c001d0d99

File: packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/types.ts
Comment (lgobbi-atix) [reply packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/types.ts]: Done in 70dd1e23cbb08257e051305e215dc03c001d0d99

File: packages/cardano-services/src/Program/loadHttpServer.ts
Comment (lgobbi-atix) [reply packages/cardano-services/src/Program/loadHttpServer.ts]: Done in 70dd1e23cbb08257e051305e215dc03c001d0d99

File: packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap
Comment (lgobbi-atix) [reply packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap:143]: We also found another issue with the order of the queries' results, apparently it was not always the same. For example, running `yarn test chainhistoryhttpservice` would pass, but running `yarn test` would fail because it was expecting the same data but in a different order in the snapshot. I added `order by` to the queries and now it's working

File: packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.ts
Comment (lgobbi-atix) [reply packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.ts]: Done in 70dd1e23cbb08257e051305e215dc03c001d0d99

File: packages/cardano-services/src/ChainHistory/ChainHistoryHttpService.ts
Comment (lgobbi-atix) [reply packages/cardano-services/src/ChainHistory/ChainHistoryHttpService.ts]: Done in 70dd1e23cbb08257e051305e215dc03c001d0d99

File: packages/cardano-services/package.json
Comment (lgobbi-atix) [reply packages/cardano-services/package.json]: Done in 70dd1e23cbb08257e051305e215dc03c001d0d99

File: packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap
Comment (lgobbi-atix) [reply packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap:143]: That should be the reason why this was happening

> Snapshots failed [here](https://github.com/input-output-hk/cardano-js-sdk/runs/6507593263?check_suite_focus=true) too after merge, but the PR checks were green

Comment (lgobbi-atix) [reply packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap:143]: CI now is failing because of dates, it looks like an issue with timezones. There is a 3 hour difference between expected and the snapshot, which I guess makes sense because I'm in UTC-3

**Review by rhyslbw — CHANGES_REQUESTED** [review]
@lgobbi-atix the changes in 70dd1e23cbb08257e051305e215dc03c001d0d99 look good, but you won't be able to squash this into the relevant commits without keeping them separate here for the review. Remember, the end goal is to merge atomic changes of the `master` branch state, and the PR branch is your staging area to do so.

File: packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/mappers.ts
Comment (rhyslbw) [reply packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/mappers.ts:217]: Yes, a separate task. I've made a ticket ADP-1792

File: packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap
Comment (rhyslbw) [reply packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap:143]: Do we have relative dates in the dataset? The DB just stores UTC. We could strip some irrelevant data like DB IDs and dates?

Comment (lgobbi-atix) [reply packages/cardano-services/test/StakePoolSeach/DbSyncStakePoolSearchProvider/__snapshots__/StakePoolSearchBuilder.test.ts.snap:143]: It looks like the issue is that db-sync stores timestamps in UTC but without timezone (`timestamp` instead of `timestamptz`), so when making a query it assumes it's stored in the computer's timezone and tries to convert it to UTC, in my case by adding 3 extra hours.

For example, for [block 1598507](https://explorer.cardano-testnet.iohkdev.io/en/block?id=7a48b034645f51743550bbaf81f8a14771e58856e031eb63844738ca8ad72298), with time `2020/07/29 03:58:16 UTC`

If I just do `SELECT time FROM block WHERE block_no = 1598507;` I get as the `db.query` result: `time:  2020-07-29T06:58:16.000Z` (which would actually be UTC+3).
But if I specify in which timezone the `time` is by doing `SELECT time AT TIME ZONE 'UTC' FROM block WHERE block_no = 1598507;` it should always return it as `time:  2020-07-29T03:58:16.000Z`, which is correct

**Review by ssong-iohk — CHANGES_REQUESTED** [review]
1. I've commented with missing test cases in `ChainHistoryBuilder.test.ts`. 
2. Unit test for `DbSyncChainHistoryProvider` class is missing. If relevant tests are existed, please point them.

File: packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts
Comment (rhyslbw) [reply packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts:37]: This would fail in the `Cardano.TransactionId` function, which is tested [here](https://github.com/input-output-hk/cardano-js-sdk/blob/dc33f153288255b256453209433f08dae1a22291/packages/core/test/Cardano/types/Transaction.test.ts). It would be best to branch off _master_, and include it there.

File: packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts
Comment (lgobbi-atix) [reply packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts:97]: You are right, changed it to `BigInt` https://github.com/input-output-hk/cardano-js-sdk/pull/243/commits/6fc21e8ed6f3d5f7a7b13c644a379001f15c9568#diff-3e16e6fa3a0c30bc7dc482538653e7edca22402ca67614ba46757d93f533b0e8L62

Comment (rhyslbw) [reply packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts:97]: Great, so with that change it's covered by type safety.

File: packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts
Comment (ssong-iohk) [reply packages/cardano-services/test/ChainHistory/DbSyncChainHistoryProvider/ChainHistoryBuilder.test.ts:37]: LGTM

**Review by ssong-iohk — APPROVED** [review]

**Review by rhyslbw — CHANGES_REQUESTED** [review]
For consistency, class filenames should match the class name. Also, the structure of the module could be improved by dropping `Provider` from the dir name:

```
packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/DbSyncChainHistory.ts
```

```
packages/cardano-services/src/ChainHistory/DbSyncChainHistory/DbSyncChainHistoryProvider.ts
```

**Comment by lgobbi-atix** [comment]
> 2. Unit test for `DbSyncChainHistoryProvider` class is missing. If relevant tests are existed, please point them.

I could add a new `DbSyncChainHistoryProvider.test.ts` file but the provider should already be covered with these kind of tests https://github.com/input-output-hk/cardano-js-sdk/pull/243/files#diff-cc6fbcd6f5f3a6dee2c12a90216ba559ea451cb187699b838f8bed6b600d65c3R101

**Comment by lgobbi-atix** [comment]
> For consistency, class filenames should match the class name. Also, the structure of the module could be improved by dropping `Provider` from the dir name:
> 
> ```
> packages/cardano-services/src/ChainHistory/DbSyncChainHistoryProvider/DbSyncChainHistory.ts
> ```
> 
> ```
> packages/cardano-services/src/ChainHistory/DbSyncChainHistory/DbSyncChainHistoryProvider.ts
> ```

@rhyslbw Done in https://github.com/input-output-hk/cardano-js-sdk/pull/243/commits/790b52abc46b7b238a27ea16ed17c50f810fea38

**Comment by rhyslbw** [comment]
> but the provider should already be covered with these kind of tests

Yes, given we're mapping 1:1, the coverage will be sufficient.

**Comment by ssong-iohk** [comment]
> > 2. Unit test for `DbSyncChainHistoryProvider` class is missing. If relevant tests are existed, please point them.
> 
> I could add a new `DbSyncChainHistoryProvider.test.ts` file but the provider should already be covered with these kind of tests https://github.com/input-output-hk/cardano-js-sdk/pull/243/files#diff-cc6fbcd6f5f3a6dee2c12a90216ba559ea451cb187699b838f8bed6b600d65c3R101

Thanks. I can see generic happy/negative cases in `chainHistoryByProvider.test.ts`. However, some negative cases could be caused by given arguments are missing or exception occurred inside the method (`transactionsByHashes()`). I will create a ticket to cover these cases later.

**Review by rhyslbw — APPROVED** [review]
Approved. Ready for final squashing, including the last into 0146f7c29148fd6430531efc196aea3f568c72d2 if that can be done cleanly.

**Review by rhyslbw — APPROVED** [review]
Great work @lgobbi-atix 🎉

**Pull request merged**
